ports/mbedTLS/azure_iot_jws_mbedtls.c

```diff
@@ -242,45 +243,39 @@ static AzureIoTResult_t prvJWS_RS256Verify( uint8_t * pucInput,
 
     if( lMbedTLSResult != 0 )
     {
-        AZLogError( ( "[JWS] mbedtls_rsa_check_pubkey res: %i", lMbedTLSResult ) );
+        AZLogError( ( "[JWS] mbedtls_rsa_check_pubkey res: %i", ( int ) lMbedTLSResult ) );
         mbedtls_rsa_free( &ctx );
         return eAzureIoTErrorFailed;
     }
 
     /* RSA */
-    lMbedTLSResult = mbedtls_rsa_pkcs1_decrypt( &ctx, NULL, NULL, MBEDTLS_RSA_PUBLIC, &ulDecryptedLength, pucSignature, pucBuffer, azureiotjwsRSA3072_SIZE );
-
-    if( lMbedTLSResult != 0 )
-    {
-        AZLogError( ( "[JWS] mbedtls_rsa_pkcs1_decrypt res: %i", lMbedTLSResult ) );
-        mbedtls_rsa_free( &ctx );
-        return eAzureIoTErrorFailed;
-    }
-
-    mbedtls_rsa_free( &ctx );
-
     xResult = prvJWS_SHA256Calculate( pucInput, ulInputLength,
-                                      pucShaBuffer );
+                                      pucBuffer );
 
     if( xResult != eAzureIoTSuccess )
     {
         AZLogError( ( "[JWS] prvJWS_SHA256Calculate failed" ) );
         return xResult;
     }
 
-    /* TODO: remove this once we have a valid PKCS7 parser. */
-    shaMatchResult = memcmp( pucBuffer + azureiotjwsPKCS7_PAYLOAD_OFFSET, pucShaBuffer, azureiotjwsSHA256_SIZE );
+    #if MBEDTLS_VERSION_NUMBER >= 0x03000000
+        lMbedTLSResult = mbedtls_rsa_pkcs1_verify( &ctx, MBEDTLS_MD_SHA256, azureiotjwsSHA256_SIZE, pucBuffer, pucSignature );
+    #else
+        lMbedTLSResult = mbedtls_rsa_pkcs1_verify( &ctx, NULL, NULL, MBEDTLS_RSA_PUBLIC, MBEDTLS_MD_SHA256, azureiotjwsSHA256_SIZE, pucBuffer, pucSignature );
+    #endif
```

Comment:
I think this works temporarily as the PKCS7 structure which ADU uses is simple. But if they decide to make it more complex (using other signed attributes), I think this could break.

[RFC 3852 Section 5.4](https://www.rfc-editor.org/rfc/rfc3852#section-5.4)

> The message digest calculation process computes a message digest on
   either the content being signed or the content together with the
   signed attributes.

I believe PKCS1 is only the message digest so we're lucky here. 

TLDR: Can we keep the `TODO` so we can hopefully change it to PKCS7 in the future if/when they add one?

Reply:
We need to make sure the ADU team gets a feedback about this design, so they take it into consideration. Paraphrasing Cristian, I am surprised the Crypto Board did not raise this flag when they got their design reviewed.